### PR TITLE
[Merged by Bors] - feat: `ring` for `PNat`

### PR DESCRIPTION
**Pull request by digama0**
This is a (relatively low-effort) implementation of `ring` on `PNat` which works by using a typeclass to lift expressions to `Nat`, where the more usual methods apply. It is written in a generic style so it can be extended with typeclasses for other types, which extends the applicability to any type which has an injective map to a commutative semiring.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/Ring/Basic.lean [hunk 1]
```diff
@@ -1086,6 +1086,19 @@ partial def eval {u} {α : Q(Type u)} (sα : Q(CommSemiring $α))
     | _ => els
   | _, _, _ => els
 
+class CSLift (α : Type u) (β : outParam (Type u)) where
+  lift : α → β
+  inj : Function.Injective lift
```
Comment (semorrison) [inline Mathlib/Tactic/Ring/Basic.lean:1091]: Could we have a doc-string? At least telling me what `CS` stands for. :-)

**Comment by semorrison** [comment]
Also, the doc-blame linter will complain here once you merge/rebase.

File: Mathlib/Tactic/Ring/Basic.lean (hunk 1 above)
Comment (digama0) [reply Mathlib/Tactic/Ring/Basic.lean:1091]: it's supposed to mean commutative semiring, although as you can see the class itself doesn't depend on it. I don't really have good naming suggestions here...

**Comment by PatrickMassot** [comment]
Is there any way to measure whether this makes `ring` slower in the traditional cases?

**Comment by jcommelin** [comment]
I suppose that we can

!bench

**Comment by digama0** [comment]
> Is there any way to measure whether this makes `ring` slower in the traditional cases?

I don't think there is any chance of this (other than the general vagaries of code being placed in different places since stuff around it changed), because the new code only runs in response to a typeclass inference failure which would have caused `ring` to fail previously.

**Comment by robertylewis** [comment]
bors merge

**Pull request closed without merging**